pandas/_libs/tslibs/timestamps.pyx

```diff
@@ -248,10 +301,12 @@ cdef class _Timestamp(ABCTimestamp):
     def __hash__(_Timestamp self):
         if self.nanosecond:
             return hash(self.value)
+        if not (1 <= self.year <= 9999):
+            # out of bounds for pydatetime
+            return hash(self.value)
```

Comment:
In these cases, `datetime.__hash__` uses the dummy 1970/1972 instead of the correct year

Reply:
My original concern was that `Timestamp(9999, 1)` would have different hashes in 1.4.x vs 1.5, but I realized that concern is not valid since `Timestamp(9999, ...)` is not supported in 1.4.x